wallet-api/src/Wallet/Generators/Mockchain/StateMachine.hs

```diff
@@ -0,0 +1,279 @@
+invariant gm = case gm^.userInvariant of
+    Nothing -> Nothing
+    Just f  -> Just $ \case
+        Initial _ -> Top
+        Concrete es a -> f es a
+
+postcondition :: State a Concrete -> Command a Concrete -> Response Concrete -> Logic
+postcondition (Concrete _ a) (UserAction Action {..}) (C es) = uaPostcondition es a
+postcondition _ _ _                                          = Top
+
+transition :: State a u -> Command a u -> Response u -> State a u
+transition (Initial gm) Initialise S             = Symbolic gm $ initialMockchain gm
+transition (Initial gm) Initialise (C es)        = Concrete es $ gm ^. initialUserState
+transition (Concrete _ a) (UserAction ua) (C es) = Concrete es (uaConcrete ua a)
+transition (Concrete _ a) _ (C es)               = Concrete es a
+transition (Symbolic gm mc) (Transaction _ tx) S = Symbolic gm $ submitTxn mc tx
+transition (Symbolic gm mc) Block S              = Symbolic gm $ validatePending mc
+transition (Symbolic gm mc) (UserAction ua) S    = foldl' (\st cmd -> transition st cmd S) s' (uaCommands ua)
+    where s' = Symbolic (over userActions tail gm) mc
+-- The following transitions can never appear due to the state machine precondition
+transition (Initial _) _ _  = error "Internal error: non-initial transition from initial state"
+transition _ Initialise _   = error "Internal error: initial transition from non-initial state"
+
+precondition :: State a Symbolic -> Command a Symbolic -> Logic
+precondition (Initial _) Initialise = Top
+precondition (Initial _) _          = Bot
+precondition _ Initialise           = Bot
+precondition _ _                    = Top
+
+generator :: State a Symbolic -> Maybe (Gen (Command a Symbolic))
+generator (Initial _)      = Just $ pure Initialise
+generator (Symbolic gm mc) = Just $ oneof $ block
+                             -- We need to make sure there are at least some entries in the UTXO set to generate
+                             -- transactions
+                             : if not $ Map.null $ mc^.utxo
+                                then [ validTransaction, invalidTransaction ] ++ userAction
+                                else []
+    where validTransaction =
+              Transaction <$> pickWallet gm <*> genValidTransaction keys keys mc (gm^.feeEstimator)
+          invalidTransaction =
+              Transaction <$> pickWallet gm <*> genInvalidTransaction keys keys mc (gm^.feeEstimator)
+          block = pure Block
+          userAction = case gm^.userActions of
+              []    -> []
+              act:_ -> [ UserAction <$> act mc ]
+          keys = Set.toList $ gm^.transactionKeys
+
+pickWallet :: GeneratorModel a -> Gen Wallet
+pickWallet = elements . wallets
+
+shrinker :: State a Symbolic -> Command a Symbolic -> [Command a Symbolic]
+shrinker _ Initialise        = []
+shrinker _ (UserAction _)    = []
+shrinker _ (Transaction _ _) = []
+shrinker _ Block             = []
```

Comment:
Does this mean we don't get shrinking?

Reply:
Not quite - you can't not get shrinking with quickcheck-state-machine, this just lets it know that the only kind of shrinking it can do is to delete an action wholesale (it can't, e.g. remove an input from a transaction).